pandas/core/internals/construction.py

```diff
@@ -240,18 +241,35 @@ def init_ndarray(values, index, columns, dtype: Optional[DtypeObj], copy: bool):
         if values.ndim == 2 and values.shape[0] != 1:
             # transpose and separate blocks
 
-            dvals_list = [maybe_infer_to_datetimelike(row) for row in values]
-            for n in range(len(dvals_list)):
-                if isinstance(dvals_list[n], np.ndarray):
-                    dvals_list[n] = dvals_list[n].reshape(1, -1)
+            dvals_list = (maybe_infer_to_datetimelike(row) for row in values)
 
             from pandas.core.internals.blocks import make_block
 
-            # TODO: What about re-joining object columns?
-            block_values = [
-                make_block(dvals_list[n], placement=[n], ndim=2)
-                for n in range(len(dvals_list))
-            ]
+            i = 0
```

Comment:
yeah this is not great. we *already* do this type of dtype grouping *inside* the block manager constructors. much better to do there. However this might be non-trivial and/or might have a bit impact performance.

so i would try to re-use one of the existing grouping rountines here instead.

Reply:
Could you point me to the module?

Reply:
i dont think we have anything quite like this, in part because i wrote something similar in #33597

Reply:
what if you dont do the make_block calls here, but instead keep construct the arrays, and then instead of calling create_block_manager_from_blocks at the end could call create_block_manager_from_arrays?

Reply:
That solves it but has a ripple effect of making other stuff fail. That constructor might need to be adjusted for use in instantiating a DataFrame.
![Screenshot 2021-01-20 at 02 14 22](https://user-images.githubusercontent.com/24586651/105117965-5bcab300-5ac5-11eb-8449-ee574c0696b3.png)

Reply:
can you post one of the actual tracebacks?

Reply:
Some of these were a missing ValueError being raised in `_form_blocks`, but it still ended up copying everything – that's just how `create_block_manager_from_arrays` is implemented. I've implemented a new `_from_array` constructor, which will falls back to `_form_blocks` in the case of a datetime conversion, but otherwise doesn't copy.

Reply:
I tried to avoid adding a kwarg to the constructor, but it seemed like that would get messier.